core/market/src/negotiation/requestor.rs

```diff
@@ -391,6 +395,52 @@ impl RequestorBroker {
             AgreementState::Terminated => AgreementStateError::Terminated(agreement.id),
         })?
     }
+
+    pub async fn terminate_agreement(&self, id: Identity, 
+        agreement_id: &AgreementId,
+    ) -> Result<(), AgreementError> {
+        let dao = self.common.db.as_dao::<AgreementDao>();
+        let mut agreement = match dao
+            .select(
+                agreement_id,
+                Some(id.identity.clone()),
+                Utc::now().naive_utc(),
+            )
+            .await
+            .map_err(|e| AgreementError::Get(agreement_id.clone(), e))?
+        {
+            None => return Err(AgreementError::NotFound(agreement_id.clone())),
+            Some(agreement) => agreement,
+        };
+
+        Err(match agreement.state {
```

Comment:
Just a note for future: We have such match on states in many places in code. Maybe we should refactor it somehow to remove duplication?